hls-plugin-api/src/Ide/Plugin/Error.hs

```diff
@@ -0,0 +1,163 @@
+pluginResponseM handler act =
+    runExceptT act >>= \case
+        Right r  -> pure $ Right r
+        Left err -> handler err
+
+-- See Note [PluginError]
+data PluginError
+  = -- |PluginInternalError should be used if something has gone horribly wrong.
+    -- All uncaught exceptions will be caught and converted to this error.
+    --
+    -- This error will be logged individually with Error, and will be converted
+    -- into an InternalError response code. It takes the highest precedence (1)
+    -- in being returned as a response to the client.
+    PluginInternalError T.Text
+    -- |PluginInvalidParams should be used if the parameters of the request are
+    -- invalid. This error means that there is a bug in the client's code
+    -- (otherwise they wouldn't be sending you requests with invalid
+    -- parameters).
+    --
+    -- This error will be logged individually with Warning, and will be
+    -- converted into a InvalidParams response code. It takes medium precedence
+    --  (2)in being returned as a response to the client.
+  | PluginInvalidParams T.Text
+    -- |PluginInvalidRequest should be used if the request is invalid. This
+    -- error means that there is a bug in the client's code (otherwise they
+    -- wouldn't be sending you an invalid request).
+    --
+    -- This error will be logged individually with Warning, and will be
+    -- converted into a InvalidRequest response code. It takes medium precedence
+    --  (2) in being returned as a response to the client.
+  | PluginInvalidRequest T.Text
+    -- |PluginParseError should be used sparingly for parse errors. Prefer to
+    -- use PluginInternalError for bugs in the plugin's code, or
+    -- PluginInvalidRequest/PluginInvalidParams for invalid requests from the
+    -- client.
+    --
+    -- This error will be logged individually with Warning, and will be
+    -- converted into a ParseError response code. It takes a medium precedence
+    --  (2) in being returned as a response to the client.
+  | PluginParseError T.Text
+    -- |PluginDependencyFailed should be thrown when a function that your plugin
+    -- depends on fails. This should only be used when the function fails
+    -- because the files the user is working on is in an invalid state.
+    --
+    -- This error takes the name of the function that failed. Prefer to catch
+    -- this error as close to the source as possible.
+    --
+    -- This error will be logged together with other errors of the same type
+    -- with Info, and will be converted into a ContentModified response. It
+    -- takes a low precedence (3) in being returned as a response to the client.
+  | PluginDependencyFailed T.Text
+    -- |PluginRequestRefused allows your handler to inspect a request before
+    -- rejecting it. In effect it allows your plugin to act make a secondary
+    -- `pluginEnabled` decision after receiving the request. This should only be
```

Comment:
dumb question - should we just replace the use of `pluginEnabled` with this?

Reply:
I put this Error in specifically for the scenario where someone wants to have more than one resolve hander for the same method with different data types. This error allows us to refuse to answer the ones that don't decode to the type provided by the handler signature.

Reply:
Right, but we _could_ use it to do the stuff that we now do with `pluginEnabled`, which would give us more options perhaps?

Reply:
I haven't looked at it extensively, but I imagine that we would lose some nice things we get by having pluginEnabled per method now.